source/extensions/tracers/skywalking/trace_segment_reporter.cc

```diff
@@ -0,0 +1,167 @@
+#include "extensions/tracers/skywalking/trace_segment_reporter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace SkyWalking {
+
+namespace {
+
+const Http::LowerCaseString& authenticationTokenKey() {
+  CONSTRUCT_ON_FIRST_USE(Http::LowerCaseString, "Authentication");
+}
+
+// Convert SegmentContext to SegmentObject.
+TraceSegmentPtr toSegmentObject(const SegmentContext& segment_context) {
+  auto new_segment_ptr = std::make_unique<SegmentObject>();
+  SegmentObject& segment_object = *new_segment_ptr;
+
+  segment_object.set_traceid(segment_context.traceId());
+  segment_object.set_tracesegmentid(segment_context.traceSegmentId());
+  segment_object.set_service(segment_context.service());
+  segment_object.set_serviceinstance(segment_context.serviceInstance());
+
+  for (const auto& span_store : segment_context.spanList()) {
+    if (!span_store->sampled()) {
+      continue;
+    }
+    auto* span = segment_object.mutable_spans()->Add();
+
+    span->set_spanlayer(SpanLayer::Http);
+    span->set_spantype(span_store->isEntrySpan() ? SpanType::Entry : SpanType::Exit);
+    // Please check
+    // https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml
+    // get more information.
+    span->set_componentid(9000);
+
+    if (!span_store->peerAddress().empty() && span_store->isEntrySpan()) {
+      span->set_peer(span_store->peerAddress());
+    }
+
+    span->set_spanid(span_store->spanId());
+    span->set_parentspanid(span_store->parentSpanId());
+
+    span->set_starttime(span_store->startTime());
+    span->set_endtime(span_store->endTime());
+
+    span->set_iserror(span_store->isError());
+
+    span->set_operationname(span_store->operation());
+
+    auto& tags = *span->mutable_tags();
+    tags.Reserve(span_store->tags().size());
+
+    for (auto& span_tag : span_store->tags()) {
+      tags.Add(std::move(const_cast<Tag&>(span_tag)));
```

Comment:
the std::move and const_cast doesn't do anything here and this still involves copy. Don't use `const_cast`.